gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java

```diff
@@ -266,6 +288,35 @@ private InsertObjectRequest buildInsertRequest(
       return requestBuilder.build();
     }
 
+    // Handles the case when a writeOffset of data read previously is being processed.
+    // This happens if a transient failure happens while uploading, and can be resumed by
+    // querying the current committed offset.
+    private InsertObjectRequest buildRequestFromCachedDataChunk(
+        TreeMap<Long, ByteString> dataChunkMap, long writeOffset) throws IOException {
+      // Resume will only work if the first request builder in the cache carries an offset
+      // not greater than the current writeOffset.
+      InsertObjectRequest request = null;
+      if (dataChunkMap.size() > 0 && dataChunkMap.firstKey() <= writeOffset) {
+        for (Map.Entry<Long, ByteString> entry : dataChunkMap.entrySet()) {
+          if (entry.getKey() + entry.getValue().size() > writeOffset) {
+            Long writeOffsetToResume = entry.getKey();
+            ByteString chunkData = entry.getValue();
+            request = buildInsertRequest(writeOffsetToResume, chunkData);
+            break;
+          }
+        }
+      }
+      if (request == null) {
+        throw new IOException(
```

Comment:
If this happens the exception will propagate all the way back to the caller. An alternative would be to catch it somewhere in this class and retry from the beginning. That would make uploads more robust for this case (although hopefully it's a very uncommon case).

I guess we could just wait and see if it's a problem before doing that, though.

Reply:
+1 do it until we actual see these kind of exception. 

I think even we do see this happens not uncommon in our tests, 1st thing we could try is to increase the buffer size instead of retrying all over again. The unretriable kindof errors need to be handled by client side anyway.